File: ForgeHub/backend/src/main/java/id/jakforge/forgehub/api/ProcessesApi.java
package id.jakforge.forgehub.api;

import id.jakforge.forgehub.data.Db;
import id.jakforge.forgehub.security.CurrentUser;
import id.jakforge.forgehub.security.ForgeHubPrincipal;
import org.springframework.http.ContentDisposition;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Proses dan paket.
 *
 * <p>Studio memanggil {@code GET /api/packages} untuk menghitung nomor versi
 * berikutnya, {@code POST /api/packages} untuk menerbitkan, dan
 * {@code GET /api/processes} untuk daftar pilihan. JakRunner mengunduh isinya
 * lewat {@code /content}.
 */
@RestController
@RequestMapping("/api")
public class ProcessesApi {

    /**
     * Batas ukuran paket.
     *
     * <p>Ada karena isinya dikirim sebagai base64 DI DALAM badan JSON: paket 64
     * MB menjadi sekitar 85 MB teks yang harus muat di memori sekaligus. Tanpa
     * batas, satu penerbitan yang keliru menjatuhkan layanan untuk semua orang.
     */
    private static final long MAKS_PAKET_BITA = 64L * 1024 * 1024;

    private final Db db;

    public ProcessesApi(Db db) {
        this.db = db;
    }

    // -----------------------------------------------------------------
    // Proses
    // -----------------------------------------------------------------

    @GetMapping("/processes")
    public List<Map<String, Object>> proses() {
        return db.rows("""
                SELECT p.id, p.name, p.package_name, p.package_version, p.environment,
                       p.description, p.created_at,
                       (SELECT count(*) FROM jobs j
                         WHERE j.tenant_id = p.tenant_id AND j.process_name = p.name) AS job_count,
                       (SELECT max(j.created_at) FROM jobs j
                         WHERE j.tenant_id = p.tenant_id AND j.process_name = p.name) AS last_run_at
                  FROM processes p
                 WHERE p.tenant_id = ?
                 ORDER BY p.name
                """, CurrentUser.get().tenantId());
    }

    /**
     * Nama yang sudah ada DIPERBARUI, bukan ditolak.
     *
     * <p>Menerbitkan ulang versi yang lebih baru adalah hal yang paling sering
     * dilakukan, dan menolaknya sebagai "sudah ada" memaksa orang menghapus
     * dulu — yang berarti sesaat prosesnya tidak ada sama sekali.
     */
    @PostMapping("/processes")
    public ResponseEntity<?> buatProses(@RequestBody(required = false) Map<String, Object> body) {
        String name = Badan.nama(body, "name");

        if (name == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Nama proses wajib diisi."));
        }

        UUID tenantId = CurrentUser.get().tenantId();

        Object adaId = db.scalar(
                "SELECT id FROM processes WHERE tenant_id = ? AND name = ?", tenantId, name);

        if (adaId != null) {
            db.exec("""
                    UPDATE processes
                       SET package_name = COALESCE(?, package_name),
                           package_version = COALESCE(?, package_version),
                           environment = COALESCE(?, environment),
                           description = COALESCE(?, description)
                     WHERE tenant_id = ? AND name = ?
                    """, Badan.teks(body, "packageName"), Badan.teks(body, "packageVersion"),
                    Badan.teks(body, "environment"), Badan.teks(body, "description"), tenantId, name);
        } else {
            db.exec("""
                    INSERT INTO processes
                        (id, tenant_id, name, package_name, package_version, environment, description, created_at)
                    VALUES (?, ?, ?, ?, ?, ?, ?, now())
                    """, Db.newId(), tenantId, name,
                    Badan.teks(body, "packageName"), Badan.teks(body, "packageVersion"),
                    Badan.teks(body, "environment", "Production"), Badan.teks(body, "description"));
        }

        return ResponseEntity.ok(Map.of("ok", true));
    }

    @DeleteMapping("/processes/{name}")
    public ResponseEntity<?> hapusProses(@PathVariable String name) {
        int terhapus = db.exec("DELETE FROM processes WHERE tenant_id = ? AND name = ?",
                CurrentUser.get().tenantId(), name);

        return terhapus == 0
                ? ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("error", "Proses '" + name + "' tidak ada."))
                : ResponseEntity.ok(Map.of("ok", true));
    }

    // -----------------------------------------------------------------
    // Paket
    // -----------------------------------------------------------------

    /**
     * Daftar paket TANPA isinya.
     *
     * <p>Kolom content sengaja tidak diambil: satu paket belasan kilobita
     * dikalikan seluruh riwayat versi akan menyeret seluruh gudang tiap kali
     * halaman dibuka, dan tidak ada satu pun yang menampilkannya.
     */
    @GetMapping("/packages")
    public List<Map<String, Object>> paket() {
        return db.rows("""
                SELECT id, name, version, description, entry_point, published_by, published_at, size_bytes
                  FROM packages
                 WHERE tenant_id = ?
                 ORDER BY name, published_at DESC
                """, CurrentUser.get().tenantId());
    }

    /**
     * Penerbitan dari Studio.
     *
     * <p>Isi paket dikirim sebagai base64 di dalam JSON, bukan multipart.
     * Penerbitnya adalah Studio di .NET Framework 4.6.2 yang menyusun
     * permintaan multipart dengan tangan; base64 lebih besar 33% tapi hanya
     * butuh satu jalan yang sama dengan permintaan lainnya.
     */
    @PostMapping("/packages")
    @Transactional
    public ResponseEntity<?> terbitkan(@RequestBody(required = false) Map<String, Object> body) {
        String name = Badan.nama(body, "name");
        String version = Badan.teks(body, "version", "1.0.0");

        if (name == null) {
            return ResponseEntity.badRequest().body(Map.of("error", "Nama paket wajib diisi."));
        }

        byte[] isi = null;
        String base64 = Badan.teks(body, "contentBase64");

        if (base64 != null && !base64.isEmpty()) {
            try {
                isi = Base64.getDecoder().decode(base64);
            } catch (IllegalArgumentException e) {
                return ResponseEntity.badRequest()
                        .body(Map.of("error", "contentBase64 bukan base64 yang sah."));
            }

            if (isi.length > MAKS_PAKET_BITA) {
                return ResponseEntity.badRequest().body(Map.of(
                        "error", "Paket terlalu besar. Batasnya " + (MAKS_PAKET_BITA / 1024 / 1024) + " MB."));
            }
        }

        ForgeHubPrincipal p = CurrentUser.get();
        UUID tenantId = p.tenantId();
        String penerbit = p.username();
        long ukuran = isi == null ? 0 : isi.length;

        Object adaId = db.scalar("""
                SELECT id FROM packages WHERE tenant_id = ? AND name = ? AND version = ?
                """, tenantId, name, version);

        if (adaId != null) {
            // COALESCE pada content: penerbitan ulang yang hanya memperbarui
            // keterangan tidak boleh MENGHAPUS isi paket yang sudah ada.
            db.exec("""
                    UPDATE packages
                       SET description = ?, entry_point = ?, published_by = ?,
                           published_at = now(), size_bytes = ?,
                           content = COALESCE(?, content)
                     WHERE tenant_id = ? AND name = ? AND version = ?
                    """, Badan.teks(body, "description"), Badan.teks(body, "entryPoint"),
                    penerbit, ukuran, isi, tenantId, name, version);
        } else {
            db.exec("""
                    INSERT INTO packages
                        (id, tenant_id, name, version, description, entry_point,
                         published_by, published_at, size_bytes, content)
                    VALUES (?, ?, ?, ?, ?, ?, ?, now(), ?, ?)
                    """, Db.newId(), tenantId, name, version, Badan.teks(body, "description"),
                    Badan.teks(body, "entryPoint"), penerbit, ukuran, isi);
        }

        // Menerbitkan paket hampir selalu berarti ingin proses dengan nama yang
        // sama tersedia untuk dijalankan. Membuatnya di sini menghemat satu
        // langkah yang mudah terlupa — dan yang terlupa itu baru terasa saat
        // pekerjaan ditolak dengan "proses belum diterbitkan".
        boolean prosesAda = db.exists(
                "SELECT count(*) FROM processes WHERE tenant_id = ? AND name = ?", tenantId, name);

        if (prosesAda) {
            db.exec("""
                    UPDATE processes SET package_name = ?, package_version = ?
                     WHERE tenant_id = ? AND name = ?
                    """, name, version, tenantId, name);
        } else {
            db.exec("""
                    INSERT INTO processes
                        (id, tenant_id, name, package_name, package_version, environment, description, created_at)
                    VALUES (?, ?, ?, ?, ?, ?, ?, now())
                    """, Db.newId(), tenantId, name, name, version,
                    Badan.teks(body, "environment", "Production"), Badan.teks(body, "description"));
        }

        Peringatan.catat(db, tenantId, "Info", "Paket diterbitkan",
                name + " " + version + " diterbitkan oleh " + (penerbit == null ? "?" : penerbit) + ".",
                "packages");

        db.exec("""
                INSERT INTO logs (tenant_id, level, message, process_name, logged_at)
                VALUES (?, 'INFO', ?, ?, now())
                """, tenantId, "Paket " + name + " " + version + " diterbitkan.", name);

        Map<String, Object> hasil = new LinkedHashMap<>();
        hasil.put("ok", true);
        hasil.put("name", name);
        hasil.put("version", version);
        hasil.put("sizeBytes", ukuran);

        return ResponseEntity.ok(hasil);
    }

    @GetMapping("/packages/{name}/{version}/content")
    public ResponseEntity<?> isiPaket(@PathVariable String name, @PathVariable String version) {
        // Isi diambil lewat kueri tersendiri yang mengembalikan bita mentah.
        // Db.rows sengaja mengubah BYTEA menjadi ukurannya, karena isi paket
        // tidak pernah benar dikirim sebagai bagian dari JSON.
        List<byte[]> isi = db.jdbc().query(
                "SELECT content FROM packages WHERE tenant_id = ? AND name = ? AND version = ?",
                (rs, i) -> rs.getBytes(1),
                CurrentUser.get().tenantId(), name, version);

        if (isi.isEmpty() || isi.get(0) == null) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND)
                    .body(Map.of("error", "Paket tidak ada atau tanpa isi."));
        }

        return ResponseEntity.ok()
                .contentType(MediaType.parseMediaType("application/zip"))
                .header(HttpHeaders.CONTENT_DISPOSITION,
                        ContentDisposition.attachment()
                                .filename(name + "." + version + ".zip")
                                .build().toString())
                .body(isi.get(0));
    }

    @DeleteMapping("/packages/{name}/{version}")
    public ResponseEntity<?> hapusPaket(@PathVariable String name, @PathVariable String version) {
        int terhapus = db.exec(
                "DELETE FROM packages WHERE tenant_id = ? AND name = ? AND version = ?",
                CurrentUser.get().tenantId(), name, version);

        return terhapus == 0
                ? ResponseEntity.status(HttpStatus.NOT_FOUND).body(Map.of("error", "Paket tidak ada."))
                : ResponseEntity.ok(Map.of("ok", true));
    }
}

File: ForgeHub/server/Api/ProcessesApi.cs
using ForgeHub.Data;

namespace ForgeHub.Api;

public static class ProcessesApi
{
    /// <summary>Batas ukuran satu paket yang diterbitkan: 64 MB.</summary>
    private const long MaxPackageBytes = 64L * 1024 * 1024;

    public static void MapProcesses(this WebApplication app)
    {
        MapPackages(app);

        // --------------------------------------------------------------
        app.MapGet("/api/processes", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var rows = Sql.Rows(connection,
                @"SELECT p.id, p.name, p.package_name, p.package_version, p.environment,
                         p.description, p.created_at,
                         (SELECT COUNT(*) FROM jobs j
                           WHERE j.tenant_id = p.tenant_id AND j.process_name = p.name) AS job_count,
                         (SELECT MAX(j.created_at) FROM jobs j
                           WHERE j.tenant_id = p.tenant_id AND j.process_name = p.name) AS last_run_at
                    FROM processes p
                   WHERE p.tenant_id = @p0
                   ORDER BY p.name",
                tenantId);

            return Results.Ok(rows);
        });

        // --------------------------------------------------------------
        // Dipakai Studio saat menerbitkan, dan dasbor saat membuat proses baru.
        // Nama yang sudah ada DIPERBARUI, bukan ditolak: menerbitkan ulang versi
        // yang lebih baru adalah hal yang paling sering dilakukan.
        app.MapPost("/api/processes", async (HttpContext context) =>
        {
            var body = await context.Body();
            var name = body.Str("name");

            if (string.IsNullOrWhiteSpace(name))
                return Results.BadRequest(new { error = "Nama proses wajib diisi." });

            using var connection = context.Db(out var tenantId);

            var exists = Sql.Count(connection,
                "SELECT COUNT(*) FROM processes WHERE tenant_id = @p0 AND name = @p1", tenantId, name) > 0;

            if (exists)
            {
                Sql.Exec(connection,
                    @"UPDATE processes
                         SET package_name = COALESCE(@p0, package_name),
                             package_version = COALESCE(@p1, package_version),
                             environment = COALESCE(@p2, environment),
                             description = COALESCE(@p3, description)
                       WHERE tenant_id = @p4 AND name = @p5",
                    body.Str("packageName"), body.Str("packageVersion"),
                    body.Str("environment"), body.Str("description"), tenantId, name);
            }
            else
            {
                Sql.Exec(connection,
                    @"INSERT INTO processes
                        (id, tenant_id, name, package_name, package_version, environment, description, created_at)
                      VALUES (@p0, @p1, @p2, @p3, @p4, @p5, @p6, @p7)",
                    Sql.NewId(), tenantId, name, body.Str("packageName"),
                    body.Str("packageVersion"), body.Str("environment", "Production"),
                    body.Str("description"), Sql.Now());
            }

            return Results.Ok(new { ok = true, created = !exists });
        });

        // --------------------------------------------------------------
        app.MapDelete("/api/processes/{name}", (HttpContext context, string name) =>
        {
            using var connection = context.Db(out var tenantId);

            var removed = Sql.Exec(connection,
                "DELETE FROM processes WHERE tenant_id = @p0 AND name = @p1", tenantId, name);

            return removed == 0
                ? Results.NotFound(new { error = "Proses '" + name + "' tidak ada." })
                : Results.Ok(new { ok = true });
        });
    }

    // ------------------------------------------------------------------

    private static void MapPackages(WebApplication app)
    {
        // Isi paketnya TIDAK ikut dalam daftar. Satu paket bisa puluhan megabita,
        // dan daftar yang membawanya akan menyeret seluruh gudang tiap kali
        // halaman dibuka.
        app.MapGet("/api/packages", (HttpContext context) =>
        {
            using var connection = context.Db(out var tenantId);

            var rows = Sql.Rows(connection,
                @"SELECT id, name, version, description, entry_point, published_by, published_at, size_bytes
                    FROM packages
                   WHERE tenant_id = @p0
                   ORDER BY name, published_at DESC",
                tenantId);

            return Results.Ok(rows);
        });

        // --------------------------------------------------------------
        // Penerbitan dari Studio.
        //
        // Isi paket dikirim sebagai base64 di dalam JSON, bukan multipart.
        // Penerbitnya adalah Studio di .NET Framework 4.6.2, yang menyusun
        // permintaan multipart dengan tangan; JSON base64 lebih besar 33% tapi
        // hanya butuh satu jalan yang sama dengan permintaan lainnya.
        app.MapPost("/api/packages", async (HttpContext context) =>
        {
            var body = await context.Body();

            var name = body.Str("name");
            var version = body.Str("version", "1.0.0");

            if (string.IsNullOrWhiteSpace(name))
                return Results.BadRequest(new { error = "Nama paket wajib diisi." });

            byte[] content = null;
            var contentBase64 = body.Str("contentBase64");

            if (!string.IsNullOrEmpty(contentBase64))
            {
                try
                {
                    content = Convert.FromBase64String(contentBase64);
                }
                catch (FormatException)
                {
                    return Results.BadRequest(new { error = "contentBase64 bukan base64 yang sah." });
                }

                if (content.LongLength > MaxPackageBytes)
                    return Results.BadRequest(new
                    {
                        error = "Paket terlalu besar. Batasnya " + (MaxPackageBytes / 1024 / 1024) + " MB.",
                    });
            }

            using var connection = context.Db(out var tenantId);

            var existingId = (string)Sql.Scalar(connection,
                "SELECT id FROM packages WHERE tenant_id = @p0 AND name = @p1 AND version = @p2",
                tenantId, name, version);

            var now = Sql.Now();
            var publisher = context.User()?.Username;
            var size = content?.LongLength ?? 0;

            if (existingId != null)
            {
                Sql.Exec(connection,
                    @"UPDATE packages
                         SET description = @p0, entry_point = @p1, published_by = @p2,
                             published_at = @p3, size_bytes = @p4,
                             content = COALESCE(@p5, content)
                       WHERE id = @p6",
                    body.Str("description"), body.Str("entryPoint"), publisher, now, size,
                    (object)content, existingId);
            }
            else
            {
                Sql.Exec(connection,
                    @"INSERT INTO packages
                        (id, tenant_id, name, version, description, entry_point, published_by, published_at, size_bytes, content)
                      VALUES (@p0, @p1, @p2, @p3, @p4, @p5, @p6, @p7, @p8, @p9)",
                    Sql.NewId(), tenantId, name, version, body.Str("description"),
                    body.Str("entryPoint"), publisher, now, size, (object)content);
            }

            // Menerbitkan paket hampir selalu berarti ingin proses dengan nama
            // yang sama tersedia untuk dijalankan. Membuatnya di sini menghemat
            // satu langkah yang mudah terlupa.
            var processExists = Sql.Count(connection,
                "SELECT COUNT(*) FROM processes WHERE tenant_id = @p0 AND name = @p1", tenantId, name) > 0;

            if (processExists)
            {
                Sql.Exec(connection,
                    "UPDATE processes SET package_name = @p0, package_version = @p1 WHERE tenant_id = @p2 AND name = @p3",
                    name, version, tenantId, name);
            }
            else
            {
                Sql.Exec(connection,
                    @"INSERT INTO processes
                        (id, tenant_id, name, package_name, package_version, environment, description, created_at)
                      VALUES (@p0, @p1, @p2, @p3, @p4, @p5, @p6, @p7)",
                    Sql.NewId(), tenantId, name, name, version,
                    body.Str("environment", "Production"), body.Str("description"), now);
            }

            ApiSupport.Alert(connection, tenantId, "Info",
                "Paket diterbitkan",
                name + " " + version + " diterbitkan oleh " + (publisher ?? "?") + ".", "packages");

            Sql.Exec(connection,
                @"INSERT INTO logs (tenant_id, level, message, process_name, logged_at)
                  VALUES (@p0, 'INFO', @p1, @p2, @p3)",
                tenantId, "Paket " + name + " " + version + " diterbitkan.", name, now);

            return Results.Ok(new { ok = true, name, version, sizeBytes = size });
        });

        // --------------------------------------------------------------
        app.MapGet("/api/packages/{name}/{version}/content", (HttpContext context, string name, string version) =>
        {
            using var connection = context.Db(out var tenantId);

            var row = Sql.Row(connection,
                "SELECT content, size_bytes FROM packages WHERE tenant_id = @p0 AND name = @p1 AND version = @p2",
                tenantId, name, version);

            if (row == null || row["content"] == null)
                return Results.NotFound(new { error = "Paket tidak ada atau tanpa isi." });

            return Results.File((byte[])row["content"], "application/zip", name + "." + version + ".zip");
        });

        // --------------------------------------------------------------
        app.MapDelete("/api/packages/{name}/{version}", (HttpContext context, string name, string version) =>
        {
            using var connection = context.Db(out var tenantId);

            var removed = Sql.Exec(connection,
                "DELETE FROM packages WHERE tenant_id = @p0 AND name = @p1 AND version = @p2",
                tenantId, name, version);

            return removed == 0
                ? Results.NotFound(new { error = "Paket tidak ada." })
                : Results.Ok(new { ok = true });
        });
    }
}
